### PR TITLE
Add unit tests for solver stanza preferences.

### PR DESCRIPTION
**Pull request by grayjay**
I added some unit tests while trying to reproduce a bug.

I also had to change the solver DSL so that it no longer enables all packages' tests by default.  I added a function (mkTestEnableAllTests) to enable tests in the existing unit tests that relied on that behavior.

### CODE REVIEW

File: cabal-install/tests/UnitTests/Distribution/Solver/Modular/Solver.hs [hunk 1]
```diff
@@ -285,7 +309,19 @@ mkTestExtLangPC :: Maybe [Extension]
                 -> [String]
                 -> SolverResult
                 -> SolverTest
-mkTestExtLangPC exts langs pkgConfigDb db label targets result = SolverTest {
+mkTestExtLangPC = mkTestAllOptions (EnableAllTests False)
```
Comment (ezyang) [inline cabal-install/tests/UnitTests/Distribution/Solver/Modular/Solver.hs]: A little sad this isn't just a combinator on top of the test.

**Comment by ezyang** [comment]
Looks OK. (I hope the test DSL isn't due for a rewrite soon haha ;)

File: cabal-install/tests/UnitTests/Distribution/Solver/Modular/Solver.hs (hunk 1 above)
Comment (grayjay) [inline cabal-install/tests/UnitTests/Distribution/Solver/Modular/Solver.hs]: You're right, that's a lot cleaner!

File: cabal-install/tests/UnitTests/Distribution/Solver/Modular/DSL.hs [hunk 2]
```diff
@@ -467,9 +471,10 @@ exResolve :: ExampleDb
           -> EnableBackjumping
           -> Maybe [ExampleVar]
           -> [ExPreference]
+          -> EnableAllTests
           -> Progress String String CI.SolverInstallPlan.SolverInstallPlan
 exResolve db exts langs pkgConfigDb targets solver mbj indepGoals reorder
-          enableBj vars prefs
+          enableBj vars prefs (EnableAllTests enableAllTests)
```
Comment (23Skidoo) [inline cabal-install/tests/UnitTests/Distribution/Solver/Modular/DSL.hs]: Maybe make `EnableAllTests` an instance of `BooleanFlag` and use `asBool`?

**Comment by 23Skidoo** [comment]
LGTM.

**Comment by 23Skidoo** [comment]
Merged, thanks!

**Comment by grayjay** [comment]
Thanks!

**Pull request merged**